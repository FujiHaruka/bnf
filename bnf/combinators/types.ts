import type { LiteralTokenNode, NamedTokenNode, TokenNode } from "../token.ts";
import type { Result } from "../utils/Result.ts";

export type Parser = (text: string, position: number) => Result<TokenNode>;
export type LiteralTokenParser = (
  text: string,
  position: number,
) => Result<LiteralTokenNode>;
export type NamedTokenParser = (
  text: string,
  position: number,
) => Result<NamedTokenNode>;
