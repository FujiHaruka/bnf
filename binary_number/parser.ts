import {
  ScanContext,
  TokenNode,
  TokenType,
  UnexpectedTokenError,
} from "./core.ts";

export const TokenTypes = {
  "binary-number": new TokenType("binary-number"),
  "binary-sequence": new TokenType("binary-sequence"),
  "binary-digit": new TokenType("binary-digit"),
  "$literal": new TokenType("$literal"),
} as const;

type ParserAtom = (ctx: ScanContext) => TokenNode | null;

function expect(ctx: ScanContext, fn: ParserAtom): TokenNode {
  const nodeOrNull = fn(ctx);
  if (nodeOrNull) {
    return nodeOrNull;
  } else {
    throw new UnexpectedTokenError(ctx);
  }
}

export const Parser = {
  /**
   * `<binary-number> ::= <binary-digit> | "1" <binary-sequence>`
   */
  "binary-number"(ctx: ScanContext): TokenNode | null {
    const startAt = ctx.cursor;

    // if it starts with a specific characotr, it can match the latter case.
    if (ctx.currentChar === "1") {
      const literalNode = new TokenNode({
        type: TokenTypes.$literal,
        children: [],
        startAt: ctx.cursor,
        endAt: ctx.cursor + 1,
      });
      ctx.nextCursor();
      const node = Parser["binary-sequence"](ctx);
      if (node) {
        return new TokenNode({
          type: TokenTypes["binary-number"],
          children: [literalNode, node],
          startAt,
          endAt: startAt + 2,
        });
      }

      // Reset cursor position
      ctx.prevCursor();
    }

    const node = Parser["binary-digit"](ctx);
    if (node) {
      ctx.nextCursor();
      return new TokenNode({
        type: TokenTypes["binary-number"],
        children: [node],
        startAt: startAt,
        endAt: startAt + 1,
      });
    } else {
      return null;
    }
  },

  /**
   * `<binary-sequence> ::= <binary-digit> | <binary-digit> <binary-sequence>`
   */
  "binary-sequence"(ctx: ScanContext): TokenNode | null {
    const startAt = ctx.cursor;
    const children: TokenNode[] = [];

    let node: TokenNode | null = null;
    do {
      node = Parser["binary-digit"](ctx);
      if (node) {
        children.push(node);
      }
    } while (node);

    if (children.length === 0) {
      return null;
    } else {
      return new TokenNode({
        type: TokenTypes["binary-sequence"],
        children,
        startAt,
        endAt: startAt + children.length,
      });
    }
  },

  /**
   * `<binary-digit> ::= "0" | "1"`
   */
  "binary-digit"(ctx: ScanContext): TokenNode | null {
    const startAt = ctx.cursor;
    switch (ctx.currentChar) {
      case "0":
      case "1": {
        ctx.nextCursor();
        return new TokenNode({
          type: TokenTypes["binary-digit"],
          children: [],
          startAt,
          endAt: startAt + 1,
        });
      }
      default:
        return null;
    }
  },
};

export function parse(text: string): TokenNode {
  const ctx = new ScanContext(text);
  const node = expect(ctx, Parser["binary-number"]);
  if (!ctx.scanFinished) {
    throw new UnexpectedTokenError(ctx);
  }

  return node;
}
