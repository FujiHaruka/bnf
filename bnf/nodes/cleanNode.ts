import { NamedTokenNode } from "../token.ts";
import { cleanupTempTokenNodes } from "./cleanupTempTokenNodes.ts";
import { flattenRecursiveNodes } from "./flattenRecursiveNodes.ts";
import { removeEmptyNodes } from "./removeEmptyNodes.ts";

export function cleanNode(node: NamedTokenNode): NamedTokenNode {
  node = cleanupTempTokenNodes(node);
  node = flattenRecursiveNodes(node);
  node = removeEmptyNodes(node);
  return node;
}
