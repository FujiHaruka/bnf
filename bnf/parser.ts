import { cleanupTempTokenNodes } from "./combinators/cleanupTempTokenNodes.ts";
import { concat } from "./combinators/concat.ts";
import { flattenRecursiveNodes } from "./combinators/flattenRecursiveNodes.ts";
import { literalTokenParser } from "./combinators/literalTokenParser.ts";
import { or } from "./combinators/or.ts";
import { NamedTokenNode, TempTokenType, TokenType } from "./token.ts";
import { UnexpectedTokenError } from "./utils/errors.ts";
import { Result } from "./utils/Result.ts";

export type RuleName =
  | "binary-number"
  | "binary-decimal"
  | "binary-integer"
  | "binary-natural-number"
  | "binary-sequence"
  | "binary-digit";

export const TokenTypes: Readonly<Record<RuleName, TokenType>> = {
  "binary-number": new TokenType("binary-number"),
  "binary-decimal": new TokenType("binary-decimal"),
  "binary-integer": new TokenType("binary-integer"),
  "binary-natural-number": new TokenType("binary-natural-number"),
  "binary-sequence": new TokenType("binary-sequence"),
  "binary-digit": new TokenType("binary-digit"),
};

export const Parser = {
  /**
   * <binary-number> ::= <binary-integer> | <binary-decimal>
   */
  "binary-number"(text: string, position: number): Result<NamedTokenNode> {
    return or(
      TokenTypes["binary-number"],
      [
        Parser["binary-integer"],
        Parser["binary-decimal"],
      ],
    )(text, position);
  },

  /**
   * <binary-decimal> ::= <binary-integer> "." <binary-sequence>
   */
  "binary-decimal"(text: string, position: number): Result<NamedTokenNode> {
    const parse = concat(
      TokenTypes["binary-decimal"],
      [
        Parser["binary-integer"],
        literalTokenParser("."),
        Parser["binary-sequence"],
      ],
    );
    return parse(text, position);
  },

  /**
   * <binary-integer> ::= "0" | "-" <binary-natural-number> | <binary-natural-number>
   */
  "binary-integer"(text: string, position: number): Result<NamedTokenNode> {
    return or(TokenTypes["binary-integer"], [
      literalTokenParser("0"),
      concat(TempTokenType, [
        literalTokenParser("-"),
        Parser["binary-natural-number"],
      ]),
      Parser["binary-natural-number"],
    ])(text, position);
  },

  /**
   * `<binary-natural-number> ::= <binary-digit> | "1" <binary-sequence>`
   */
  "binary-natural-number"(
    text: string,
    position: number,
  ): Result<NamedTokenNode> {
    return or(TokenTypes["binary-natural-number"], [
      Parser["binary-digit"],
      concat(
        TempTokenType,
        [
          literalTokenParser("1"),
          Parser["binary-sequence"],
        ],
      ),
    ])(text, position);
  },

  /**
   * `<binary-sequence> ::= <binary-digit> | <binary-digit> <binary-sequence>`
   */
  "binary-sequence"(text: string, position: number): Result<NamedTokenNode> {
    return or(TokenTypes["binary-sequence"], [
      Parser["binary-digit"],
      concat(TempTokenType, [
        Parser["binary-digit"],
        Parser["binary-sequence"],
      ]),
    ])(text, position);
  },

  /**
   * `<binary-digit> ::= "0" | "1"`
   */
  "binary-digit"(text: string, position: number): Result<NamedTokenNode> {
    return or(
      TokenTypes["binary-digit"],
      [
        literalTokenParser("0"),
        literalTokenParser("1"),
      ],
    )(text, position);
  },
};

export function parse(text: string): NamedTokenNode {
  const node = Parser["binary-number"](text, 0)
    .map<NamedTokenNode>(cleanupTempTokenNodes)
    .map<NamedTokenNode>(flattenRecursiveNodes)
    .unwrap();

  if (node.endAt !== text.length) {
    throw new UnexpectedTokenError({
      ruleName: "$root",
      char: text.charAt(node.endAt),
      position: node.endAt,
    });
  }

  return node;
}
