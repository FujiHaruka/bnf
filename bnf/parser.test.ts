import {
  assert,
  assertObjectMatch,
  assertSnapshot,
  describe,
  it,
  makeloc,
  posix,
} from "../deps-test.ts";
import { cleanNode } from "./nodes/cleanNode.ts";
import { NamedTokenParser } from "./combinators/types.ts";
import { parse, Parser } from "./parser.ts";

describe("Parsers", () => {
  const tests: [NamedTokenParser, string, Record<string, unknown>][] = [
    // <rule-char> ::= <letter> | <digit> | "-"
    [Parser["rule-char"], "-", {
      type: "rule-char",
      children: [{
        value: "-",
      }],
    }],
    [Parser["rule-char"], "A", {
      type: "rule-char",
      children: [{
        type: "letter",
        children: [{
          value: "A",
        }],
      }],
    }],
    [Parser["rule-char"], "0", {
      type: "rule-char",
      children: [{
        type: "digit",
        children: [{
          value: "0",
        }],
      }],
    }],
    // <rule-name> ::= <letter> | <rule-name> <rule-char>
    [Parser["rule-name"], "fo-ba", {
      type: "rule-name",
      startAt: 0,
      endAt: 5,
      children: [{
        type: "letter",
      }, {
        type: "rule-char",
      }, {
        type: "rule-char",
      }, {
        type: "rule-char",
      }, {
        type: "rule-char",
      }],
    }],
    // <literal> ::= '"' <text1> '"' | "'" <text2> "'"
    [Parser["literal"], '"A"', {
      type: "literal",
      startAt: 0,
      endAt: 3,
      children: [{
        value: '"',
      }, {
        type: "text1",
      }, {
        value: '"',
      }],
    }],
    [Parser["literal"], '""', {
      type: "literal",
      startAt: 0,
      endAt: 2,
      children: [{
        value: '"',
      }, {
        value: '"',
      }],
    }],
    // <term> ::= <literal> | "<" <rule-name> ">"
    [Parser["term"], "'A'", {
      type: "term",
      startAt: 0,
      endAt: 3,
      children: [{
        type: "literal",
        children: [{
          value: "'",
        }, {
          type: "text2",
        }, {
          value: "'",
        }],
      }],
    }],
    [Parser["term"], "<ro-ba>", {
      type: "term",
      startAt: 0,
      endAt: 7,
      children: [{
        value: "<",
      }, {
        type: "rule-name",
      }, {
        value: ">",
      }],
    }],
    [Parser["term"], `"'"`, {
      type: "term",
      startAt: 0,
      endAt: 3,
      children: [{
        type: "literal",
        children: [{
          value: '"',
        }, {
          type: "text1",
        }, {
          value: '"',
        }],
      }],
    }],
    // <list> ::= <term> | <term> <opt-whitespace> <list>
    [Parser["list"], "<term> <opt-whitespace> <list>", {
      type: "list",
      startAt: 0,
      endAt: 30,
    }],
    // <line-end> ::= <opt-whitespace> <EOL> | <line-end> <line-end>
    [Parser["line-end"], " \n\n  \n", {
      type: "line-end",
      startAt: 0,
      endAt: 6,
    }],
    [Parser["line-end"], "\n", {
      type: "line-end",
      startAt: 0,
      endAt: 1,
    }],
    // <opt-whitespace> ::= " " <opt-whitespace> | ""
    [Parser["opt-whitespace"], " ", {
      type: "opt-whitespace",
      startAt: 0,
      endAt: 1,
      children: [{
        value: " ",
      }],
    }],
    // <rule> ::= <opt-whitespace> "<" <rule-name> ">" <opt-whitespace> "::=" <opt-whitespace> <expression> <line-end>
    [
      Parser["rule"],
      '<rule> ::= <opt-whitespace> "<" <rule-name> ">" <opt-whitespace> "::=" <opt-whitespace> <expression> <line-end>\n',
      {
        startAt: 0,
        endAt: 112,
      },
    ],
  ];

  tests.forEach(([parse, text, expected]) => {
    it(`${parse.name} parses \`${text.replaceAll("\n", "\\n")}\``, () => {
      const result = parse(text, 0)
        .map(cleanNode);
      if (parse.name === "opt-whitespace") {
        console.log(result.unwrap().toJSON());
      }
      assert(result.isOk());
      assertObjectMatch(result.unwrap().toJSON(), expected);
    });
  });
});

describe(parse.name, () => {
  const fixtures = [
    "binary_number.bnf",
    // "bnf.bnf",
  ];

  fixtures.forEach((fixture) => {
    Deno.test(`parse / success with ${fixture}`, async (t) => {
      const { __dirname } = makeloc(import.meta);
      const filepath = posix.join(__dirname, "__fixtures__", fixture);
      const bnf = Deno.readTextFileSync(filepath);
      await assertSnapshot(t, parse(bnf));
      // Deno.writeTextFileSync(`${fixture}.json`, JSON.stringify(parse(bnf), null, 2))
    });
  });
});
