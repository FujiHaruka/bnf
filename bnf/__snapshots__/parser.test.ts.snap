export const snapshot = {};

snapshot[`parse / success "0" 1`] = `
NamedTokenNode {
  children: [
    NamedTokenNode {
      children: [
        LiteralTokenNode {
          endAt: 1,
          startAt: 0,
          type: TokenType {
            name: "\$literal",
          },
          value: "0",
        },
      ],
      endAt: 1,
      startAt: 0,
      type: TokenType {
        name: "binary-integer",
      },
    },
  ],
  endAt: 1,
  startAt: 0,
  type: TokenType {
    name: "binary-number",
  },
}
`;

snapshot[`parse / success "1" 1`] = `
NamedTokenNode {
  children: [
    NamedTokenNode {
      children: [
        NamedTokenNode {
          children: [
            NamedTokenNode {
              children: [
                LiteralTokenNode {
                  endAt: 1,
                  startAt: 0,
                  type: TokenType {
                    name: "\$literal",
                  },
                  value: "1",
                },
              ],
              endAt: 1,
              startAt: 0,
              type: TokenType {
                name: "binary-digit",
              },
            },
          ],
          endAt: 1,
          startAt: 0,
          type: TokenType {
            name: "binary-natural-number",
          },
        },
      ],
      endAt: 1,
      startAt: 0,
      type: TokenType {
        name: "binary-integer",
      },
    },
  ],
  endAt: 1,
  startAt: 0,
  type: TokenType {
    name: "binary-number",
  },
}
`;

snapshot[`parse / success "10" 1`] = `
NamedTokenNode {
  children: [
    NamedTokenNode {
      children: [
        NamedTokenNode {
          children: [
            LiteralTokenNode {
              endAt: 1,
              startAt: 0,
              type: TokenType {
                name: "\$literal",
              },
              value: "1",
            },
            NamedTokenNode {
              children: [
                NamedTokenNode {
                  children: [
                    LiteralTokenNode {
                      endAt: 2,
                      startAt: 1,
                      type: TokenType {
                        name: "\$literal",
                      },
                      value: "0",
                    },
                  ],
                  endAt: 2,
                  startAt: 1,
                  type: TokenType {
                    name: "binary-digit",
                  },
                },
              ],
              endAt: 2,
              startAt: 1,
              type: TokenType {
                name: "binary-sequence",
              },
            },
          ],
          endAt: 2,
          startAt: 0,
          type: TokenType {
            name: "binary-natural-number",
          },
        },
      ],
      endAt: 2,
      startAt: 0,
      type: TokenType {
        name: "binary-integer",
      },
    },
  ],
  endAt: 2,
  startAt: 0,
  type: TokenType {
    name: "binary-number",
  },
}
`;

snapshot[`parse / success "11" 1`] = `
NamedTokenNode {
  children: [
    NamedTokenNode {
      children: [
        NamedTokenNode {
          children: [
            LiteralTokenNode {
              endAt: 1,
              startAt: 0,
              type: TokenType {
                name: "\$literal",
              },
              value: "1",
            },
            NamedTokenNode {
              children: [
                NamedTokenNode {
                  children: [
                    LiteralTokenNode {
                      endAt: 2,
                      startAt: 1,
                      type: TokenType {
                        name: "\$literal",
                      },
                      value: "1",
                    },
                  ],
                  endAt: 2,
                  startAt: 1,
                  type: TokenType {
                    name: "binary-digit",
                  },
                },
              ],
              endAt: 2,
              startAt: 1,
              type: TokenType {
                name: "binary-sequence",
              },
            },
          ],
          endAt: 2,
          startAt: 0,
          type: TokenType {
            name: "binary-natural-number",
          },
        },
      ],
      endAt: 2,
      startAt: 0,
      type: TokenType {
        name: "binary-integer",
      },
    },
  ],
  endAt: 2,
  startAt: 0,
  type: TokenType {
    name: "binary-number",
  },
}
`;

snapshot[`parse / success "100" 1`] = `
NamedTokenNode {
  children: [
    NamedTokenNode {
      children: [
        NamedTokenNode {
          children: [
            LiteralTokenNode {
              endAt: 1,
              startAt: 0,
              type: TokenType {
                name: "\$literal",
              },
              value: "1",
            },
            NamedTokenNode {
              children: [
                NamedTokenNode {
                  children: [
                    LiteralTokenNode {
                      endAt: 2,
                      startAt: 1,
                      type: TokenType {
                        name: "\$literal",
                      },
                      value: "0",
                    },
                  ],
                  endAt: 2,
                  startAt: 1,
                  type: TokenType {
                    name: "binary-digit",
                  },
                },
                NamedTokenNode {
                  children: [
                    LiteralTokenNode {
                      endAt: 3,
                      startAt: 2,
                      type: TokenType {
                        name: "\$literal",
                      },
                      value: "0",
                    },
                  ],
                  endAt: 3,
                  startAt: 2,
                  type: TokenType {
                    name: "binary-digit",
                  },
                },
              ],
              endAt: 3,
              startAt: 1,
              type: TokenType {
                name: "binary-sequence",
              },
            },
          ],
          endAt: 3,
          startAt: 0,
          type: TokenType {
            name: "binary-natural-number",
          },
        },
      ],
      endAt: 3,
      startAt: 0,
      type: TokenType {
        name: "binary-integer",
      },
    },
  ],
  endAt: 3,
  startAt: 0,
  type: TokenType {
    name: "binary-number",
  },
}
`;

snapshot[`parse / success "101" 1`] = `
NamedTokenNode {
  children: [
    NamedTokenNode {
      children: [
        NamedTokenNode {
          children: [
            LiteralTokenNode {
              endAt: 1,
              startAt: 0,
              type: TokenType {
                name: "\$literal",
              },
              value: "1",
            },
            NamedTokenNode {
              children: [
                NamedTokenNode {
                  children: [
                    LiteralTokenNode {
                      endAt: 2,
                      startAt: 1,
                      type: TokenType {
                        name: "\$literal",
                      },
                      value: "0",
                    },
                  ],
                  endAt: 2,
                  startAt: 1,
                  type: TokenType {
                    name: "binary-digit",
                  },
                },
                NamedTokenNode {
                  children: [
                    LiteralTokenNode {
                      endAt: 3,
                      startAt: 2,
                      type: TokenType {
                        name: "\$literal",
                      },
                      value: "1",
                    },
                  ],
                  endAt: 3,
                  startAt: 2,
                  type: TokenType {
                    name: "binary-digit",
                  },
                },
              ],
              endAt: 3,
              startAt: 1,
              type: TokenType {
                name: "binary-sequence",
              },
            },
          ],
          endAt: 3,
          startAt: 0,
          type: TokenType {
            name: "binary-natural-number",
          },
        },
      ],
      endAt: 3,
      startAt: 0,
      type: TokenType {
        name: "binary-integer",
      },
    },
  ],
  endAt: 3,
  startAt: 0,
  type: TokenType {
    name: "binary-number",
  },
}
`;

snapshot[`parse / success "110" 1`] = `
NamedTokenNode {
  children: [
    NamedTokenNode {
      children: [
        NamedTokenNode {
          children: [
            LiteralTokenNode {
              endAt: 1,
              startAt: 0,
              type: TokenType {
                name: "\$literal",
              },
              value: "1",
            },
            NamedTokenNode {
              children: [
                NamedTokenNode {
                  children: [
                    LiteralTokenNode {
                      endAt: 2,
                      startAt: 1,
                      type: TokenType {
                        name: "\$literal",
                      },
                      value: "1",
                    },
                  ],
                  endAt: 2,
                  startAt: 1,
                  type: TokenType {
                    name: "binary-digit",
                  },
                },
                NamedTokenNode {
                  children: [
                    LiteralTokenNode {
                      endAt: 3,
                      startAt: 2,
                      type: TokenType {
                        name: "\$literal",
                      },
                      value: "0",
                    },
                  ],
                  endAt: 3,
                  startAt: 2,
                  type: TokenType {
                    name: "binary-digit",
                  },
                },
              ],
              endAt: 3,
              startAt: 1,
              type: TokenType {
                name: "binary-sequence",
              },
            },
          ],
          endAt: 3,
          startAt: 0,
          type: TokenType {
            name: "binary-natural-number",
          },
        },
      ],
      endAt: 3,
      startAt: 0,
      type: TokenType {
        name: "binary-integer",
      },
    },
  ],
  endAt: 3,
  startAt: 0,
  type: TokenType {
    name: "binary-number",
  },
}
`;

snapshot[`parse / success "111" 1`] = `
NamedTokenNode {
  children: [
    NamedTokenNode {
      children: [
        NamedTokenNode {
          children: [
            LiteralTokenNode {
              endAt: 1,
              startAt: 0,
              type: TokenType {
                name: "\$literal",
              },
              value: "1",
            },
            NamedTokenNode {
              children: [
                NamedTokenNode {
                  children: [
                    LiteralTokenNode {
                      endAt: 2,
                      startAt: 1,
                      type: TokenType {
                        name: "\$literal",
                      },
                      value: "1",
                    },
                  ],
                  endAt: 2,
                  startAt: 1,
                  type: TokenType {
                    name: "binary-digit",
                  },
                },
                NamedTokenNode {
                  children: [
                    LiteralTokenNode {
                      endAt: 3,
                      startAt: 2,
                      type: TokenType {
                        name: "\$literal",
                      },
                      value: "1",
                    },
                  ],
                  endAt: 3,
                  startAt: 2,
                  type: TokenType {
                    name: "binary-digit",
                  },
                },
              ],
              endAt: 3,
              startAt: 1,
              type: TokenType {
                name: "binary-sequence",
              },
            },
          ],
          endAt: 3,
          startAt: 0,
          type: TokenType {
            name: "binary-natural-number",
          },
        },
      ],
      endAt: 3,
      startAt: 0,
      type: TokenType {
        name: "binary-integer",
      },
    },
  ],
  endAt: 3,
  startAt: 0,
  type: TokenType {
    name: "binary-number",
  },
}
`;

snapshot[`parse / success "1.01" 1`] = `
NamedTokenNode {
  children: [
    NamedTokenNode {
      children: [
        NamedTokenNode {
          children: [
            NamedTokenNode {
              children: [
                NamedTokenNode {
                  children: [
                    LiteralTokenNode {
                      endAt: 1,
                      startAt: 0,
                      type: TokenType {
                        name: "\$literal",
                      },
                      value: "1",
                    },
                  ],
                  endAt: 1,
                  startAt: 0,
                  type: TokenType {
                    name: "binary-digit",
                  },
                },
              ],
              endAt: 1,
              startAt: 0,
              type: TokenType {
                name: "binary-natural-number",
              },
            },
          ],
          endAt: 1,
          startAt: 0,
          type: TokenType {
            name: "binary-integer",
          },
        },
        LiteralTokenNode {
          endAt: 2,
          startAt: 1,
          type: TokenType {
            name: "\$literal",
          },
          value: ".",
        },
        NamedTokenNode {
          children: [
            NamedTokenNode {
              children: [
                LiteralTokenNode {
                  endAt: 3,
                  startAt: 2,
                  type: TokenType {
                    name: "\$literal",
                  },
                  value: "0",
                },
              ],
              endAt: 3,
              startAt: 2,
              type: TokenType {
                name: "binary-digit",
              },
            },
            NamedTokenNode {
              children: [
                LiteralTokenNode {
                  endAt: 4,
                  startAt: 3,
                  type: TokenType {
                    name: "\$literal",
                  },
                  value: "1",
                },
              ],
              endAt: 4,
              startAt: 3,
              type: TokenType {
                name: "binary-digit",
              },
            },
          ],
          endAt: 4,
          startAt: 2,
          type: TokenType {
            name: "binary-sequence",
          },
        },
      ],
      endAt: 4,
      startAt: 0,
      type: TokenType {
        name: "binary-decimal",
      },
    },
  ],
  endAt: 4,
  startAt: 0,
  type: TokenType {
    name: "binary-number",
  },
}
`;

snapshot[`parse / success "1.10" 1`] = `
NamedTokenNode {
  children: [
    NamedTokenNode {
      children: [
        NamedTokenNode {
          children: [
            NamedTokenNode {
              children: [
                NamedTokenNode {
                  children: [
                    LiteralTokenNode {
                      endAt: 1,
                      startAt: 0,
                      type: TokenType {
                        name: "\$literal",
                      },
                      value: "1",
                    },
                  ],
                  endAt: 1,
                  startAt: 0,
                  type: TokenType {
                    name: "binary-digit",
                  },
                },
              ],
              endAt: 1,
              startAt: 0,
              type: TokenType {
                name: "binary-natural-number",
              },
            },
          ],
          endAt: 1,
          startAt: 0,
          type: TokenType {
            name: "binary-integer",
          },
        },
        LiteralTokenNode {
          endAt: 2,
          startAt: 1,
          type: TokenType {
            name: "\$literal",
          },
          value: ".",
        },
        NamedTokenNode {
          children: [
            NamedTokenNode {
              children: [
                LiteralTokenNode {
                  endAt: 3,
                  startAt: 2,
                  type: TokenType {
                    name: "\$literal",
                  },
                  value: "1",
                },
              ],
              endAt: 3,
              startAt: 2,
              type: TokenType {
                name: "binary-digit",
              },
            },
            NamedTokenNode {
              children: [
                LiteralTokenNode {
                  endAt: 4,
                  startAt: 3,
                  type: TokenType {
                    name: "\$literal",
                  },
                  value: "0",
                },
              ],
              endAt: 4,
              startAt: 3,
              type: TokenType {
                name: "binary-digit",
              },
            },
          ],
          endAt: 4,
          startAt: 2,
          type: TokenType {
            name: "binary-sequence",
          },
        },
      ],
      endAt: 4,
      startAt: 0,
      type: TokenType {
        name: "binary-decimal",
      },
    },
  ],
  endAt: 4,
  startAt: 0,
  type: TokenType {
    name: "binary-number",
  },
}
`;

snapshot[`parse / success "-1" 1`] = `
NamedTokenNode {
  children: [
    NamedTokenNode {
      children: [
        LiteralTokenNode {
          endAt: 1,
          startAt: 0,
          type: TokenType {
            name: "\$literal",
          },
          value: "-",
        },
        NamedTokenNode {
          children: [
            NamedTokenNode {
              children: [
                LiteralTokenNode {
                  endAt: 2,
                  startAt: 1,
                  type: TokenType {
                    name: "\$literal",
                  },
                  value: "1",
                },
              ],
              endAt: 2,
              startAt: 1,
              type: TokenType {
                name: "binary-digit",
              },
            },
          ],
          endAt: 2,
          startAt: 1,
          type: TokenType {
            name: "binary-natural-number",
          },
        },
      ],
      endAt: 2,
      startAt: 0,
      type: TokenType {
        name: "binary-integer",
      },
    },
  ],
  endAt: 2,
  startAt: 0,
  type: TokenType {
    name: "binary-number",
  },
}
`;

snapshot[`parse / success "0" 2`] = `
NamedTokenNode {
  children: [
    NamedTokenNode {
      children: [
        LiteralTokenNode {
          endAt: 1,
          startAt: 0,
          type: TokenType {
            name: "\$literal",
          },
          value: "0",
        },
      ],
      endAt: 1,
      startAt: 0,
      type: TokenType {
        name: "binary-integer",
      },
    },
  ],
  endAt: 1,
  startAt: 0,
  type: TokenType {
    name: "binary-number",
  },
}
`;

snapshot[`parse / success "-10.10" 1`] = `
NamedTokenNode {
  children: [
    NamedTokenNode {
      children: [
        NamedTokenNode {
          children: [
            LiteralTokenNode {
              endAt: 1,
              startAt: 0,
              type: TokenType {
                name: "\$literal",
              },
              value: "-",
            },
            NamedTokenNode {
              children: [
                LiteralTokenNode {
                  endAt: 2,
                  startAt: 1,
                  type: TokenType {
                    name: "\$literal",
                  },
                  value: "1",
                },
                NamedTokenNode {
                  children: [
                    NamedTokenNode {
                      children: [
                        LiteralTokenNode {
                          endAt: 3,
                          startAt: 2,
                          type: TokenType {
                            name: "\$literal",
                          },
                          value: "0",
                        },
                      ],
                      endAt: 3,
                      startAt: 2,
                      type: TokenType {
                        name: "binary-digit",
                      },
                    },
                  ],
                  endAt: 3,
                  startAt: 2,
                  type: TokenType {
                    name: "binary-sequence",
                  },
                },
              ],
              endAt: 3,
              startAt: 1,
              type: TokenType {
                name: "binary-natural-number",
              },
            },
          ],
          endAt: 3,
          startAt: 0,
          type: TokenType {
            name: "binary-integer",
          },
        },
        LiteralTokenNode {
          endAt: 4,
          startAt: 3,
          type: TokenType {
            name: "\$literal",
          },
          value: ".",
        },
        NamedTokenNode {
          children: [
            NamedTokenNode {
              children: [
                LiteralTokenNode {
                  endAt: 5,
                  startAt: 4,
                  type: TokenType {
                    name: "\$literal",
                  },
                  value: "1",
                },
              ],
              endAt: 5,
              startAt: 4,
              type: TokenType {
                name: "binary-digit",
              },
            },
            NamedTokenNode {
              children: [
                LiteralTokenNode {
                  endAt: 6,
                  startAt: 5,
                  type: TokenType {
                    name: "\$literal",
                  },
                  value: "0",
                },
              ],
              endAt: 6,
              startAt: 5,
              type: TokenType {
                name: "binary-digit",
              },
            },
          ],
          endAt: 6,
          startAt: 4,
          type: TokenType {
            name: "binary-sequence",
          },
        },
      ],
      endAt: 6,
      startAt: 0,
      type: TokenType {
        name: "binary-decimal",
      },
    },
  ],
  endAt: 6,
  startAt: 0,
  type: TokenType {
    name: "binary-number",
  },
}
`;
